### PR TITLE
Dmstgcn

### PR DESCRIPTION
**Pull request by Kazeya27**

### CODE REVIEW

**Comment by Kazeya27** [comment]
**result**

PeMSD4        
    ----------------------------------------DMSTGCN-------libcity
  - Horizon1         
  --------------------------MAE-------0.8847-------0.8946
  -------------------------MAPE-------0.0161-------0.0163
  -------------------------RMSE-------1.6929-------1.7025
- Horizion 3
  --------------------------MAE-------1.3002-------1.3058
  -------------------------MAPE-------0.0255-------0.0254
  -------------------------RMSE-------2.7769-------2.7734
- Horizion 6
  --------------------------MAE-------1.5875-------1.5868
  -------------------------MAPE-------0.0333-------0.0331
  -------------------------RMSE-------3.6339-------3.6242
- Horizion 12
  --------------------------MAE-------1.8787-------1.8791
  -------------------------MAPE-------0.0415-------0.0410
  -------------------------RMSE-------4.3814-------4.3729

PeMSD8        
    ----------------------------------------DMSTGCN-------libcity
  - Horizon1         
  --------------------------MAE-------0.7222-------0.7282
  -------------------------MAPE-------0.0132-------0.0132
  -------------------------RMSE-------1.4217-------1.4299
- Horizion 3
  --------------------------MAE-------1.0667-------1.0803
  -------------------------MAPE-------0.0209-------0.0209
  -------------------------RMSE-------2.4405-------2.4714
- Horizion 6
  --------------------------MAE-------1.3082-------1.3271
  -------------------------MAPE-------0.0280-------0.0278
  -------------------------RMSE-------3.2861-------3.3165
- Horizion 12
  --------------------------MAE-------1.5522-------1.5656
  -------------------------MAPE-------0.0358-------0.0354
  -------------------------RMSE-------4.0522-------4.0472

**Comment by Kazeya27** [comment]
{
  "max_epoch": 200,

  "learner": "adam",
  "learning_rate": 0.001,
  "lr_epsilon": 1e-8,
  "weight_decay": 0.0001,

  "lr_patience": 10,
  "lr_decay_ratio": 0.3,
  "lr_threshold": 1e-3,
  "lr_scheduler": "reducelronplateau",

  "clip_grad_norm": true,
  "max_grad_norm": 5,
  "use_early_stop": true,
  "patience": 20
}
{
  "batch_size": 64,
  "cache_dataset": true,
  "num_workers": 0,
  "pad_with_last_sample": true,
  "train_rate": 0.6,
  "eval_rate": 0.2,
  "scaler": "standard",
  "load_external": false,
  "normal_external": true,
  "ext_scaler": "standard",
  "input_window": 12,
  "output_window": 12,
  "add_time_in_day": false,
  "add_day_in_week": false
}

File: libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                             x_train.shape[0] + x_val.shape[0] + x_test.shape[0]) % self.timeslots
+        if self.normal_external:
+            x_train[..., self.output_dim:] = self.ext_scaler.transform(x_train[..., self.output_dim:])
+            y_train[..., self.output_dim:] = self.ext_scaler.transform(y_train[..., self.output_dim:])
+            x_val[..., self.output_dim:] = self.ext_scaler.transform(x_val[..., self.output_dim:])
+            y_val[..., self.output_dim:] = self.ext_scaler.transform(y_val[..., self.output_dim:])
+            x_test[..., self.output_dim:] = self.ext_scaler.transform(x_test[..., self.output_dim:])
+            y_test[..., self.output_dim:] = self.ext_scaler.transform(y_test[..., self.output_dim:])
+
+        x_trains, y_trains, idx_trains = self.shuffle(x_train, y_train, idx_train)
```
[the remark this answers is not in the document]
Comment (Kazeya27) [reply libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py:55]: 之前没注意libcity打乱过数据，这里看源库里有打乱，就打乱一下训练集

[the remark this answers is not in the document]
Comment (aptx1231) [reply libcity/data/dataset/dataset_subclass/dmstgcn_dataset.py:55]: 为什么要打乱？

**Pull request closed without merging**